File: External References/Devart/dotConnect/Linq/Samples/LinqConnect/CS/SubmitChanges/SubmitChangesControl.cs
using System.Linq;
using System.Windows.Forms;
using System;

namespace LinqDemo.SubmitChanges {

  public partial class SubmitChangesControl : BaseControl {

    public SubmitChangesControl() : base("SubmitChanges") {
      InitializeComponent();
    }

    internal override event EventHandler UpdateChanged;

    internal override void OpenClick() {

      var companies = from c in crmDataContext.Companies
                      select c;

      dgCompany.DataSource = companies;

      base.OpenClick();
    }

    internal override void CloseClick() {

      dgCompany.CommitEdit(DataGridViewDataErrorContexts.Commit);
      dgCompany.DataSource = typeof(CrmDataContext.Company);
      base.CloseClick();
    }

    internal override void ControlsEnabled(bool state) {

      tsbUpdate.Enabled = state;
    }

    private void tsbUpdate_Click(object sender, System.EventArgs e) {

      try {
        if (crmDataContext != null) {
          crmDataContext.SubmitChanges();
          if (UpdateChanged != null)
            UpdateChanged(this, null);
        }
      }
      catch (Exception ex) {
        MessageBox.Show(ex.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
      }
    }

    private void dgCompany_DataError(object sender, System.Windows.Forms.DataGridViewDataErrorEventArgs e) {

      if (e.Exception is ArgumentException)
        MessageBox.Show("The value is not valid.", MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
      else
        MessageBox.Show(e.Exception.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
      dgCompany.CancelEdit();
    }
  }
}

File: External References/Devart/dotConnect/Linq/Samples/LinqConnect/VB/SubmitChanges/SubmitChangesControl.vb
Public Class SubmitChangesControl
  Inherits BaseControl

  Public Sub New()
    MyBase.New("SubmitChanges")
    InitializeComponent()
  End Sub

  Friend Overrides Sub OpenClick()
    Dim companies = From c In myCrmDataContext.Companies _
                      Select c

    dgCompany.DataSource = companies
    MyBase.OpenClick()
  End Sub

  Friend Overrides Sub CloseClick()
    dgCompany.CommitEdit(DataGridViewDataErrorContexts.Commit)
    dgCompany.DataSource = Nothing
    MyBase.CloseClick()
  End Sub

  Friend Overrides Sub ControlsEnabled(ByVal state As Boolean)
    Me.tsbUpdate.Enabled = state
  End Sub

  Private Sub tsbUpdate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles tsbUpdate.Click
    Try
      If Not (CrmDataContext Is Nothing) Then
        CrmDataContext.SubmitChanges()
      End If
    Catch ex As Exception
      MessageBox.Show(ex.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
    End Try
  End Sub

  Private Sub dgCompany_DataError(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DataGridViewDataErrorEventArgs) Handles dgCompany.DataError

    If TryCast(e.Exception, ArgumentException) IsNot Nothing Then
      MessageBox.Show("The value is not valid.", MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
    Else
      MessageBox.Show(e.Exception.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
    End If
    dgCompany.CancelEdit()
  End Sub
End Class
